### DISCUSSION TITLE
Jyers gets Linear

### DISCUSSION THREAD

**Discussion by ghost** [Jyers/Marlin#707] [General Discussion]
I saw the Jyers gets linear commit. is it working without mod the board?  Yours is 4.2.2?  I have both but would prefer I guess if have to mod then the 4.2.7.   anyway, had any time to play with it? is it awsome?  If mod, which mod or doc did you follow?

**Comment by Tuncay-Ayhan** [comment]
That is his own personal config, it's included as config to compile yourself. But not for public.

**Reply by ghost** [reply]
I know, I always compile myself. Jusrt wondered what difference he can see in the Linear, and if modified the board (some say klipper works without mod?) .  If any good Ill mod my 4.2.7 board, thats got the 2225 stepper drivers I think.

**Reply by Tuncay-Ayhan** [reply]
> I know, I always compile myself. Jusrt wondered what difference he can see in the Linear, and if modified the board (some say klipper works without mod?) .  If any good Ill mod my 4.2.7 board, thats got the 2225 stepper drivers I think.

That I don't know. I just wanted to point out that that was his own config

**Reply by ghost** [reply]
I cant seem to find any info on the custom icon sets tools, had a play but cant seem to generate the ICO file.  or if I do then it wont flash to display anyway.  Wiki says like coming soon for docs, hopin there is something in the discussions I missed and cant find in search

**Reply by eecue** [reply]
I posted a detailed how to, search the discussions for the icons.  That being said, I'm also compiling myself and interested in getting Linear Advance working. @Jyers teach us the way!

**Comment by Jyers** [comment]
I used this mod [E3V2 TMC UART](https://www.reddit.com/r/ender3v2/comments/n3clew/ender_3v2_422427_board_tmc_uart_mods/). I have been talking to the guy who put it together for a while and he did a really great job of it. He details how to set up UART control for all the stepper motors, but I opted to just do the extruder driver. So I just connected the SWDIO pin from next to the LCD port to the UART pin on the stepper driver.

I have also modified the pins file in the source code to make it so all UART pin configuration can be done from within the configuration.h file. For instance in my configuration file, I set up the pin with `#define TMC_UART_PIN_E PA13` (PA13 = SWDIO) and the same can be done with the other motors through `TMC_UART_PIN_X` `TMC_UART_PIN_Y` and `TMC_UART_PIN_Z` without any need to modify the pins_CREALITY_V4.h file.

Here's how the mod ended up looking:

![20210504_113512](https://user-images.githubusercontent.com/76993396/117187480-f7004380-ada9-11eb-8254-8a24f96037f7.jpg)
![20210504_113448](https://user-images.githubusercontent.com/76993396/117187560-0b444080-adaa-11eb-885b-3093afd7fbf7.jpg)
![20210504_113456](https://user-images.githubusercontent.com/76993396/117187515-ff587e80-ada9-11eb-8a2a-8c808f7640bf.jpg)

**Reply by eecue** [reply]
Awesome, saw this on Reddit this week and wasn’t sure if I needed it. Any upside to doing all 4 stepper drivers?

**Reply by gothcha** [reply]
Nice job! How did you find the performance after the mod? You opted for just the extruder, do you believe there is a big benefit to do the X/Y? I can imagine that using LA on Z is pretty pointless.

**Reply by gothcha** [reply]
I also noticed your board has the A on it which apparently signifies TMC2208 which makes sense as the LA worked!

**Reply by Jyers** [reply]
Haven't tested much, but so far it definitely gives a noticeable improvement to surface finish and corners. As far as the benefit of doing the other motors, as far as I can tell it just gives better current control, so idk how much it really improves print quality.

**Reply by eecue** [reply]
I'm also an A and I opened up and tried to solder the jumper on today but it was a no go... mostly because I just tilted my printer sideways and tried to do it in situ... I'm going to do it proper and remove the board this weekend.

**Reply by colgateam** [reply]
Did you still modify the following line even though you just soldered one of the drivers?

#define CHOPPER_TIMING CHOPPER_DEFAULT_24V

**Reply by Jyers** [reply]
@colgateam I did, haven't tried with and without, but it seems to be working with that modified

**Reply by Reggi3** [reply]
> Haven't tested much, but so far it definitely gives a noticeable improvement to surface finish and corners. As far as the benefit of doing the other motors, as far as I can tell it just gives better current control, so idk how much it really improves print quality.

My understanding is that you get better torque, it uses less current and the motors respond quicker, this is useful for higher speeds.  I haven't tested them myself yet but I had spoken to some of the people on marlin discord to get this information, with the person I spoke to using all motors in spread mode.  I'm guessing it's a bit noisier, especially as the motors start hitting musical tones at around 100mm/s and above.

**Reply by tiagofreire-pt** [reply]
> 
> 
> @colgateam I did, haven't tried with and without, but it seems to be working with that modified

This option was merged a long time ago: https://github.com/MarlinFirmware/Marlin/pull/13565/files

Why is the 12V as the default value still?

**Reply by Dylotic** [reply]
If you do the mod and you want to reverse it. Do you need to desolder to use it the old way again or does it make it configurable both ways with the mod?

**Reply by Reggi3** [reply]
@Dylotic It's simple, just set the step mode back to stealth chop and you're back to how they used to be.

**Comment by Giadej** [comment]
I have TMC2209 drivers on my 4.2.2 board (sd slot shield marked with letter B). Is it necessary to have a soldered jumper in order to use the UART control function?

![creality4 2 2(TMC2209)](https://user-images.githubusercontent.com/80910217/117196811-431da900-ade7-11eb-86f4-76db4204d434.jpeg)

**Reply by duncio** [reply]
Hey, I never knew those letters on sd card slot actually meant something! Could you explain that a bit, please?

**Reply by Jyers** [reply]
From what I've been told, you shouldn't need a hardware mod, but don't quote me on it

**Reply by Giadej** [reply]
> 
> 
> From what I've been told, you shouldn't need a hardware mod, but don't quote me on it

Could you please direct me what changes I need to make to the firmware to enable the linear advance function?

P.S.: As mentioned above, I have a 4.2.2 with TMC2209 drivers

**Reply by Reggi3** [reply]
So here's the thing, we don't actually know which boards do the Linear Advance out of the box, so I'm having to come up with a moving theory about why it works on certain 4.2.x boards.  There are 3 possibilities:
1. The 2209 based boards are physically different and have uart enabled via hardware
2. The 2209 based boards are physically different and have the spread pin wired to enable spread cycle mode
2.  Either 2209 or A4988 drivers drivers just work with Linear Advance in stealthchop mode

All 3 modes of operation are viable but as I don't have a board with 2209 chips on it I can't trace the circuit to see whether anything changes.

So my advice on a 2209/A4988 board is to do everything except the hardware mod and uart pin/config changes (Basically enable the LA defines, disable S_CURVE, should be enough I think?).  It will either work, or it will work for a very short period of time and then the extruder motor will stop extruding.  If it doesn't work, you can of course, just revert back to your favourite Jyers firmware and be back to where you were in less than a minute.

The main issue on 2208 based boards is that in stealthchop mode, the fast pace of extruder moves causes back EMF through the TMC2208 chip, which eventually triggers an error state in the chip, causing it to shut down and the motor to stop spinning, you can re-enable the driver by toggling the EN pin, just disable steppers from the menu, then do a move on any axis to toggle the en pin, of course this is not a solution if it stops in the middle of a print.  From the information I've gathered so far, this is down to a design flaw in the TMC2208 chip, that doesn't exist in the TMC2209 or A4988 chips.

**Reply by tiagofreire-pt** [reply]
> 
> 
> > From what I've been told, you shouldn't need a hardware mod, but don't quote me on it
> 
> Could you please direct me what changes I need to make to the firmware to enable the linear advance function?
> 
> P.S.: As mentioned above, I have a 4.2.2 with TMC2209 drivers

I'm in the same situation, with my Ender 3 V2 board 4.2.2 + "B" letter written on it.

For having LA support without any hardware mods, @Jyers , could you provide some initial configuration, couldn't you? 👍

**Reply by Jyers** [reply]
@tiagofreire-pt If you aren't doing a mod, hypothetically it should be as simple as disabling s curve and enabling lin advance

**Reply by tiagofreire-pt** [reply]
@Jyers , OK. But using this:
`#define E0_DRIVER_TYPE TMC2209`

Instead of:
`//#define E0_DRIVER_TYPE TMC2208_STANDALONE`

Requires:

`Marlin\src\HAL\STM32F1\../../inc/SanityCheck.h:2563:4: error: #error "TMC2208 or TMC2209 on E0 requires E0_HARDWARE_SERIAL or E0_SERIAL_(RX|TX)_PIN."`

Which pin should I use for?

**Reply by Jyers** [reply]
@tiagofreire-pt That's gonna take some individual research, cause I have not idea, like I said mine was a 2208 so I connected the pin myself 😅

**Reply by tiagofreire-pt** [reply]
@Jyers , if I leave the E0 stepper driver as "STANDALONE" will it work?

**Reply by Jyers** [reply]
@tiagofreire-pt I really don't know, depends on which of @Reggi3's hypotheses are correct. All you can do it try. Use the test pattern from the marlin wiki to test if it is working.

**Reply by tiagofreire-pt** [reply]
@Jyers , why did you disable STEALTHCHOP_E?

**Reply by Reggi3** [reply]
disabling stealthchop enables spread cycle but only on a tmc uart enabled board.

If you want test LA without doing a mod, all you need is to define LIN_ADVANCE and comment out S_CURVE_ACCELERATION.  None of the other changes are necessary and shouldn't be done unless you're doing the uart mod.

edit:  There is also another define:
```#define CHOPPER_TIMING CHOPPER_DEFAULT_12V```
Which I think should be:
```#define CHOPPER_TIMING CHOPPER_DEFAULT_24V```
It wasn't in the original mod, I did a little research based on the fact that we run 24v motors.

**Comment by gothcha** [comment]
![4A29A913-5C98-43FD-9BA6-13F88804CB9C](https://user-images.githubusercontent.com/80574048/117205234-41ed7b80-ade9-11eb-91b0-d50b325c3ece.jpeg)

**Reply by duncio** [reply]
Thanks!

**Reply by vhristenco** [reply]
I thought all 4.2.7 boards have TMC2225, however mine has a "B" on the sdcard slot...

**Reply by Reggi3** [reply]
Me too!! This is a very interesting development...  As it stands we now have 5 4.2.x variants, using 4 diferent chips.
3x 4.2.2 boards:
1. TMC2208 trinamic chips
2. TMC2209 trinamic chips
3. A4988 chips (supposed to be allegro, reports of them being clones but not confirmed)

2x 4.2.7 boards:
1. TMC2225 trinamic chips
2. TMC2209 trinamic chips

The TMC2225 and TMC2209 chips are more interesting, as it might be possible to enable linear advance without needing to do any uart based hardware modding to enable spread cycle.  In theory all it should take is taking the spread pin high or low, so shorting that pin to gnd or an appropriate voltage., although it might mean cutting a trace on the board to separate it from other pins that are tied together.  If you refer back to my other posts in this thread, it is highly possible that the 2225 or 2209 or A4988 based boards will work with just some config changes.

The reason I say that it's highly possible are that people have been using LA out of the box on some boards, diving through all of the marlin configs and pin files for creality 4.2.x boards, there do not appear to be any configs or pin configurations that enable LA or define any UART modes for TMC drivers.  

So my working theory is that these just work with a correctly configured firmware.  Either that or creality have custom firmware and haven't released that information, my money is on it just working but I could easily be wrong.  I guess there's one way to find out :-D

**Reply by Reggi3** [reply]
> I thought all 4.2.7 boards have TMC2225, however mine has a "B" on the sdcard slot...

Could you take a picture of your mainboard and post it here please?

**Reply by Makney** [reply]
I have a 4.2.2 2208 (A) board, before I used Jyers FImrware, I can use with the original Creality firmware LA. There must be certain settings in the firmware that allow this Creality just doesn't give them out.

Pictures from a XYZ Cube
- Right Creality 1.1.1 Marlin 2.0.1
- Left Jyers 3.1.2
![20210506_204108](https://user-images.githubusercontent.com/38569334/117400468-fd560300-af02-11eb-9a06-37448d17c1a4.jpg)

**Reply by eecue** [reply]
Image isn’t working. Can you just attach directly?

**Reply by colgateam** [reply]
It's also strange the official store on AliExpress has 3 different options for ordering a silent board. I wonder what the differences are if any

https://a.aliexpress.com/_mL9xow9

**Reply by Reggi3** [reply]
At least one of those boards is a silent board, rather than a 4.2.x, the other 2, 1 is definitely a 4.2.7 and the one for the ender 3 v2, looks like a 4.2.2 board but doesn't have a version number printed on it.

**Reply by bocca86** [reply]
mine doesnt have any letter written on the sd card slot, any way to find out wich version it is?

**Reply by Reggi3** [reply]
Only way to find out is to look at the model number (4.2.7 will have 2225 or 2209 chips, which should be moddable via either the spread pin method or uart method) and if that doesn't give you any clues (4.2.2 board doesn't give any clues), remove a heatsink from one of the driver chips and clean off the thermal glue, then identify the chip directly.

**Comment by tiagofreire-pt** [comment]
Well, it seems to be working on my Ender 3 V2, 4.2.2 "B" board, with no mods:

My current configuration is available here: https://github.com/tiagofreire-pt/Marlin/tree/Ender_3_V2_Extensible_UI/Marlin

**Reply by eecue** [reply]
Is it not even worth it to try on my 4.2.2 "A" board? Does anything bad happen if it doesn't work?

**Reply by t3chguy** [reply]
Nothing bad happens, just your steppers won't move reliably

**Reply by tiagofreire-pt** [reply]
> Is it not even worth it to try on my 4.2.2 "A" board? Does anything bad happen if it doesn't work?

Yes, it's worth trying.

**Reply by msdossajo** [reply]
@tiagofreire-pt when using LA does the E stepper motor become noisier? Mine did, also have a 4.2.2 "B", after applying the mod it is quieter.

**Reply by tiagofreire-pt** [reply]
> @tiagofreire-pt when using LA does the E stepper motor become noisier? Mine did, also have a 4.2.2 "B", after applying the mod it is quieter.

Yes, I noticed it as well. 

@Reggi3, could you share any ideas about this behavior?

**Reply by Reggi3** [reply]
If you're switching from silent mode to spread cycle, it will be noisier because of the the way spread cycle works, I am not entirely sure why it goes quiet on a 2209 board if you do the spread pin mod.  There may be some other related defines that I/we are unaware of that could improve noise/behaviour.

**Reply by darco415** [reply]
Its known, Linear Advance is loud

**Reply by tiagofreire-pt** [reply]
I tried today in a "A" Creality 4.2.2 Ender 2 V2 board, without success, just with new firmware. Oddly, the extruder works if I demand it to extrude (changing filament, e.g.), but printing it does not move at all. Neither generates any noise.

I'm not sure if the cause is about the lack of hardware mod on TMC 2208 or any other variable playing at this.

**Reply by Reggi3** [reply]
If the driver chips are TMC2208, then it will need the uart mod done for the extruder to work in linear advance mode.  The most likely reason that a simple extrude works is because it's a single, uninterrupted extruder move, that doesn't actually require any intervention from the lin advance code because of a lack of X or Y movement.

**Reply by hitboxr** [reply]
@tiagofreire-pt I realize this is old but the link to your config no longer works, could you share what you changed? Obviously define LIN_ADVANCE but did you switch to spreadcycle on the E stepper in the firmware? Or were you just using LA with stealthchop? I have a B board with 2209s and I'm curious about trying this out...

**Reply by tiagofreire-pt** [reply]
> @tiagofreire-pt I realize this is old but the link to your config no longer works, could you share what you changed? Obviously define LIN_ADVANCE but did you switch to spreadcycle on the E stepper in the firmware? Or were you just using LA with stealthchop? I have a B board with 2209s and I'm curious about trying this out...

Just straight enabling `LIN_ADVANCE`.

**Comment by tiagofreire-pt** [comment]
For those whom might want to determine the right K-factor, I share 2 custom gcodes generated from [here](https://marlinfw.org/tools/lin_advance/k-factor.html), specifically for the Ender 3 V2 with bowden tube (stock + BLTouch):

[K-factor lines from 0 up to 2](https://meocloud.pt/link/e67832de-0361-4e71-8e7e-debd0861d0a6/K_LA_patternkfactor_0-2.gcode/)
[K-factor lines from 0.4 up to 0.9](https://meocloud.pt/link/d8c0f4a4-55cd-4449-a585-772904e2bace/K_LA_patternkfactor_0.4-0.9.gcode/)

More info in: https://marlinfw.org/docs/features/lin_advance.html#calibration

**Reply by Makney** [reply]
a Other Methode to kalibrate the K Value is with the Calibration-Shapes Plugin in Cura
https://github.com/5axes/Calibration-Shapes/wiki/Linear-Pressure-Advance-Tower

**Reply by tiagofreire-pt** [reply]
That cube is sliced here, ready to print (with BLTouch support/G29):

[Linear advance cube](https://meocloud.pt/link/5d70a1a1-a260-4ca6-9ef1-9d4d7db1c832/Linear%20Advanced%20Cube.gcode/)

**Comment by smooreace** [comment]
Just gotta lick it like a 9 volt. B models taste like mango

**Reply by Reggi3** [reply]
the simplest way is to remove one of the heatsinks and clean the thermal glue off and read the chip.   I believe all enders are now moving over to 32bit boards, if it's a 4.2.2 and NOT on an E3v2, then you should check the config files for your printer (Ender 3, pro etc.) for #define E0_DRIVER_TYPE and see what it says the driver is.  All 4.2.7 boards so far have moddable driver chips on them.

**Reply by msdossajo** [reply]
After removing the heatsink on my 4.2.2 "B" board I could confirm the 2209 driver. 
![IMG_1807](https://user-images.githubusercontent.com/72068472/117980906-57542f80-b32c-11eb-8b10-4e6fd3bdfc73.jpg)

**Reply by tiagofreire-pt** [reply]
@msdossajo Do you have any good resolution photos of all the board for tracing the pins of the TMC2209 and the STM32? Aiming to a complete support if the STM32 has any connected pins allowing full UART control.

**Reply by gothcha** [reply]
@tiagofreire-pt I am in the process of creating a full schematic for the 4.2.2 board. I have painstakingly reverse engineered the layout. Should be complete in a day or so. I will post a copy on here which I hope will help in understanding.
@msdossajo The photo of the 2209 is a big help!

**Reply by Reggi3** [reply]
I take it you are aware that the 2209 chip allows you to control all 4 chips via uart through a single stm32 pin and addressing?  It's in the datasheet which I'm pretty sure you're reading, only 4 chips can be chained but it makes it easier to do the whole mode with just 1 pin, so it's ideal for a stock board if you want to do extra mods too.

**Reply by tiagofreire-pt** [reply]
1 pin for all the 4 chips (2209) will led to "write access" only, not full UART mode, it seems. That feedback is on the reddit thread.

**Reply by msdossajo** [reply]
@tiagofreire-pt unfortunately I don't have a high resolution of all the board, but I can try to get it. @Reggi3 I have plans to remove the R19 and connect a resistor to another pin of the SMT and try to read/write the TMC.

**Reply by Reggi3** [reply]
> 1 pin for all the 4 chips (2209) will led to "write access" only, not full UART mode, it seems. That feedback is on the reddit thread.

I suggest a read of page 15 of the [datasheet for the TMC2209 driver chips](https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC2209_Datasheet_V103.pdf)

The TMC2209 does not have the same single wire Uart functionality, it's addressable, addresses set via MS pins, Write AND Read are available in this mode.

Could you post a link or screenshot of the post discussing 2209 and a single wire, I can't see it on my thread at all?

**Reply by tiagofreire-pt** [reply]
> 
> 
> The TMC2209 does not have the same single wire Uart functionality, it's addressable, addresses set via MS pins, Write AND Read are available in this mode.
> Could you post a link or screenshot of the post discussing 2209 and a single wire, I can't see it on my thread at all?

@Reggi3 

Here: https://drive.google.com/file/d/1oUj07YvqexBF21gjLlnnvFgyJlPtkhSL/view

In: "How to gain access to TMC UART on XYZ"


_The TMC UART mode works in various ways. The first is independently with all motor drivers
requiring their own pin on the STM32 to connect to the PDN_UART pin on each driver chip. The
second way is to connect all of the PDN_UART pins to a single pin on the STM32, but that only
gives us write access, so we'll be concentrating on the former method._

**Reply by gothcha** [reply]
@tiagofreire-pt Check out my post with the schematics #814. It doesn’t appear that there are any existing connections from the UART to the STM. But there are plenty of spare pins, but most would require soldering direct the the pins.

**Reply by gothcha** [reply]
@msdossajo You do not need to remove the resistor, the STM can easily  drive UART with the 10K load

**Reply by Reggi3** [reply]
> > The TMC2209 does not have the same single wire Uart functionality, it's addressable, addresses set via MS pins, Write AND Read are available in this mode.
> > Could you post a link or screenshot of the post discussing 2209 and a single wire, I can't see it on my thread at all?
> 
> @Reggi3
> 
> Here: https://drive.google.com/file/d/1oUj07YvqexBF21gjLlnnvFgyJlPtkhSL/view
> 
> In: "How to gain access to TMC UART on XYZ"
> 
> _The TMC UART mode works in various ways. The first is independently with all motor drivers requiring their own pin on the STM32 to connect to the PDN_UART pin on each driver chip. The second way is to connect all of the PDN_UART pins to a single pin on the STM32, but that only gives us write access, so we'll be concentrating on the former method._

@tiagofreire-pt Thanks for quoting me my own mod (I am the author, I did the research and wrote that document :-D ), I see why you thought it wasn't possible on the 2209 chips and why I couldn't see an actual discussion on the thread, however, it is possible.

Just to be clear, at the time of writing I didn't have access to a TMC2209 based board (still don't), information was fairly limited at that point, I really just wanted a single method.  In the meantime, we have had more people testing/modding their boards, where we have newer slightly different methods for some boards. 

I did know that TMC2209 should be capable of doing a single uart pin mod, it's a better organised chip than the 2208 but given that the mod is a little more fraught with issues, it wasn't something I really wanted to encourage people to attempt at that point.

I would 100% trust the trinamic TMC2209 pdf in this matter over the musings I made in the original pdf/reddit post.

I hope that clears up the confusion for you?

**Reply by Reggi3** [reply]
> @msdossajo You do not need to remove the resistor, the STM can easily drive UART with the 10K load

@msdossajo this is correct, it should work fine with R19 still attached.

**Reply by gothcha** [reply]
@msdossajo do you still have the heat sink off the 2209? There is something I would like to verify for my schematic.

**Reply by msdossajo** [reply]
@gothcha no, but I can take it off. Just let me know what pictures do you want. 
Here are some pictures with lower quality I have.

![2021-05-09-120325](https://user-images.githubusercontent.com/72068472/118256420-585c9c80-b4a5-11eb-8afd-31ad8e4423bd.jpg)
![2021-05-09-120656](https://user-images.githubusercontent.com/72068472/118256423-58f53300-b4a5-11eb-9fb6-9d5b8b4208fd.jpg)
![2021-05-09-121454](https://user-images.githubusercontent.com/72068472/118256425-598dc980-b4a5-11eb-8426-5fa48aaf0188.jpg)
![2021-05-09-121623](https://user-images.githubusercontent.com/72068472/118256428-598dc980-b4a5-11eb-980e-241565b8baf2.jpg)

**Reply by gothcha** [reply]
@msdossajo Thanks this is great I can clearly see that pin 7 is not going anywhere.

**Reply by Reggi3** [reply]
I had pin 7 marked as tied to whatever makes it stealthy (gnd/low?), I can't imagine the havoc having that thing floating would cause :-D  This has still got me wondering which board is the 'LA out of the box 4.2.x board'?

**Reply by t3chguy** [reply]
The datasheet explicitly says it may be left unconnected, so that implies there's a pull and the pin is not high impedance.

![image](https://user-images.githubusercontent.com/2403652/118277864-2eb16e80-b4c1-11eb-9a62-ca730101039b.png)

**Reply by gothcha** [reply]
This @t3chguy is an interesting question, it is clear that the SPREAD pin is not wired on the 4.2.2 board, so even if one has a TMC2209 chip there is no way that the firmware can put it into SpreadCycle mode using the pin. It is possible with the hardware mod to connect the UART, but I cannot see how it can properly run LA 'out of the box' as people have reported. It may be possible in the firmware to enable LA mode, but it would be running in StealthChop mode which is not recommended!

Also the 4.2.7 boards Creality in their infinite wisdom have tied the UART to the DIAG pin which stops Serial mode from operating so overall without modifications none of the boards are likely to be successful.

@Reggi3 Yes the data sheet states that Pin 7 may be left unconnected (it has an internal pulldown) which enables StealthChop.

**Reply by Reggi3** [reply]
Ahh, thanks for the update, always happy to gain new knowledge :-)  WRT 2209 and LA out of the box, this is purely my own speculation, the 2208 chips specifically have a fault in stealthchop mode that causes them to error which shows up when you try to use LA in stealthchop mode, afaik the 2209 chips do not have the same flaw, I could be mistaken but this is the information I have.  If it's not the 2209 chip that works out of the box with LA, then it's one or both of the 4988 based boards.

**Comment by Dylotic** [comment]
Someone here modded the 4.2.7 with 2225 drivers? Really interested to know is this is possible.

**Reply by colgateam** [reply]
I tried it and it didn't seem to work.
 Still works non-linear on unchanged firmware, so doesn't ruin the board if you're willing to try.

**Reply by Reggi3** [reply]
It should work using the spread pin method:
https://github.com/Jyers/Marlin/commit/ecb49cb5685ff60832cd37c8eb81f5cb218fd877#commitcomment-50506042

Or if you want full uart, there is a method in this post (cut the trace between pin 17 and pin 18 on the TMC225 chip):
https://github.com/MarlinFirmware/Marlin/issues/16076#issuecomment-817088535

**Reply by gothcha** [reply]
I did the @Jyers UART mod to Extruder on my 2208 board and the LA works as expected. I calibrated and found Kp of 0.8 was ideal for my V2 with Bowden tube.

**Reply by Giadej** [reply]
> 
> 
> I did the @Jyers UART mod to Extruder on my 2208 board and the LA works as expected. I calibrated and found Kp of 0.8 was ideal for my V2 with Bowden tube.

Could you share your pre-compiled version of the firmware with the changes for the Linear Advance?

**Comment by gothcha** [comment]
I could however I have lots of other specific changes to the configuration that may not be suitable for your setup

**Reply by Giadej** [reply]
> 
> 
> I could however I have lots of other specific changes to the configuration that may not be suitable for your setup

Did you only make changes to the "configuration.h" file? 
In this case, could you tell me what values ​​you changed to enable the function?

**Reply by gothcha** [reply]
@Giadej You need to change a few lines in the config and adv files. Follow the link at the top and through to the pdf for the detailed info  I also followed the changes from Jyers commit
https://github.com/Jyers/Marlin/commit/ecb49cb5685ff60832cd37c8eb81f5cb218fd877
But ignored the current mods. 

I do not have access to my VSC code today so sorry this is the best I can do to help at the moment.

**Comment by darco415** [comment]
Recently received a replacement mb from creality..4.2.7/H 2225....followed above config changes...all steppers work fine via command line. Extruding during a print seems to be an issue. even tho manually extruding works fine. Running Kfactor test gcodes...at this point Im assuming this is considered a failed LA attempt, meaning a typical issue w/mb not modded?

**Reply by gothcha** [reply]
Yes without the mods the extruder works in StealthChop which is not ideal for using LA. Either wire the Spread Pin high or cut the DIAG track and connect the UART pin.

**Comment by ghost** [comment]
ok,  BaDboD goes Linear :-) 

Creality 4.2.7 controller board (TMC2225) or H.  Should not need to remove driver heatsink but will need a soldering iron

**Method 1 (Blob method)**
   Below stepper driver (viewed when reading CREALITY text right way up on the board)  remove R48.
   Short the exposed pad **closest to stepper driver** of R48 to R52, right next to and left of R48.  Small bit of wire or solder blob will do. Picture shows removing resistor. Highlighted in red is where your solder blob will go.

![blob-method](https://user-images.githubusercontent.com/7991688/120913483-abe79200-c6c1-11eb-9be4-b65c76844a2c.png)

**Method 2 (One wire method)**
   No need to remove the resistor, simply connect the two points shown in picture with a short peice of wire. Also has benifit of allowing for UART mod later without removing this mod.

![one-wire-method](https://user-images.githubusercontent.com/7991688/120913451-8195d480-c6c1-11eb-8722-0f5f52ab9beb.png)


Both ways will pull pin 19 high togglling from STEALTHCHOP to SPREADCYCLE without needing UART control.

In Marlin configuration_adv.h : 
uncomment 
   #define LIN_ADVANCE

edit line
  #define LIN_ADVANCE_K 0.22     
you could set zero and configure in slicer or set something more reasonable like 0.125 for your immediate test print gratification :-)

build and flash.
Enjoy Linear Advance in all it's Creality witheld glory. You may want to use the Marlin 'K-factor Calibration Pattern' to get a _more you_ K value.

Probably also want to disable combing and coasting in slicer, but see what you get first in case you don't need to.

 **To undo** just remove solder blob/bit of wire you added, no need to put resistor back (Blob method) as documentation states LOW or NC (No Connection) for STEALTHCHOP so it is easy to revese if you so desire.

**Reply by eecue** [reply]
Whoa! so no wires added? Which board does this work on? I have B.

**Reply by ghost** [reply]
It's the Creality 4.2.7 board. At moment of writing they all have the TMC2225 stepper driver? and mod works for all 4.2.7 boards? (according to Creality) Though some one posted they had 4.2.7 with B but Creality advertising etc all say TMC2225 so need to look at the chip to see if it is oblong or square. You should be able to see without ripping the heat sink off.  Again, according to Creality, all newly purchased 4.2.7 and new  printers have the TMC2225.  You can also check by seeing if the resistor layout looks same. The TMC2209 also has a toggle pin so it should be just as easy to mod but I dont have any to play with.

One in photo is H, not sure about one in printer now until it finshes current print.  It is 2225 so probably has H on it.

Works as pin 19 on the 2225 (pin 7 of 2209) toggles STEALTHCHOP/SPREADCYCLE, so there is no need for UART. so no wires. Correct pin is labeled SPREAD on the datasheets.

If you can get me a photo of the the board around the driver  (without heatsink) , I should be able to work out an easy way to toggle it. 

Actually,  just  a photo of the row of rersistors/caps next to vref trim pot might do. From photo in a previous message in this thread, looks like the bottom left, below-right of C33 might well be the target for removal. Check for continuity from top of that resistor to pin 7 of driver, if so remove it (the resistor).  and jumper pad nearest driver to next black 010 resistor along, top pad nearest to stepper driver. There should be 3.3v from VCC_IO on that resistor acting as pull up for DIAG same as on the TMC2225.     That picture is the 4.2.2 board though so might not be the same layout at all on the 4.2.7 ?  Thats why I need photo of it.

**Reply by darco415** [reply]
How do you remove the resistor?

**Reply by ghost** [reply]
Add a little leaded solder to each end (with soldering iron), this will lower the melting point of the factory unleaded solder. Then angle the soldering iron so you can gently touch both ends at the same time. With tweezers **gently** tap the resistor sideways and when it starts to move around, lift it (it will usually stick to your soldering iron so may not need to lift with tweezers.  At no point should you use any force at all. If the solder is not melting when touching it gently with the iron, clean your iron and get a little fresh solder on the iron, make sure your iron is ~300C temperature. All this needs **no pressure of any kind** or you WILL rip the pads off the board and game over.

You will need a microscope, digital multimeter, pin tip probes, fine tipped temperature regulated solder iron (I use TS80 plugged into power bank) fine leaded flux cored solder (or some fine solder and good flux), wet tip cleaning pad/sponge or copper brillo pad style tip cleaner.  tweezers, steady hands, some experience to get solder to melt, come off and go on without using any force or pressure on the tip. Patience.  Also if you have hot air and lots of experience then you could use that, but this board is one large heat sink so you reallly need to know what your doing if your gonna use hot air or you will just charcoal the board.

Failing that then get some one else to do it.  :-)

**Reply by ghost** [reply]
You don't have to remove the resistor, you can jumper directly from the driver side of R48 to the NON driver side of R52.   Just remember that NON driver side of R48 is 0V (GND) and NON driver side of R52 is VCC_IO (3.3v logic output from driver).  If you accidentally short these two together you risk letting out the magic smoke.  If you are careful not to do that, then a short link of wire from driver side R48 to NON driver side R52 is all you really need to do.    (Updated post with this method and picture)

**Reply by Reggi3** [reply]
I'd like to point out that a microscope is very useful but not entirely necessary, a jewellers loupe/glasses attachment is good enough, you'll just be getting up close and personal with all of your tools and the pcb :-)

**Reply by tom2199** [reply]
I've done the mod in question with my 4.2.7 board, but if I try to print the calibration pattern the extruder slips on the filament and it makes some kind of rattling noise, it prints almost mothing just the first line which is K factor 0 I guess. Do I need to up my extruder motor voltage or am I missing something here?

**Comment by smooreace** [comment]
The same way you solder the contacts together

**Reply by darco415** [reply]
wow, great advice

**Reply by CaptJocco** [reply]
I know this is an old post, but hopefully someone is still monitoring it -
Question -  after doing this mod, do I need to change the driver settings in the firmware from "Standalone" to just the regular driver? I'm very new to compiling firmware and I'm trying to setup my first personalized firmware for my Ender 3 Pro (4.2.7, Sprite Extruder Pro, BL Touch), and it's VERY difficult to learn how to do it without at least some basic understanding of all this stuff, as every site I look at to try and learn either a) explains what to do but leaves out little things, like what "standalone" means and why that should be the setting, or b) assumes you know something about coding, drivers, in depth 3D print settings...etc, and as such it flies right over my head. I wish there was a place where you could find someone to help walk you through all this.

**Reply by hitboxr** [reply]
@CaptJocco this question was already asked here: https://github.com/Jyers/Marlin/discussions/707#discussioncomment-706999 and answered here: https://github.com/Jyers/Marlin/discussions/707#discussioncomment-712009

That second comment should answer most of your questions regarding defines for testing linear advance. I'd recommend reading through that whole thread, as it discusses most of the idiosyncrasies relating to linear advance on E3V2s.

Worth noting that if you have an "A" board, linear advance almost certainly will not work without a hardware mode to enable spreadcycle due to a (presumed) hardware fault in the TMC2208s that come on those boards. Some people here have had success with "B" boards running linear advance out of the box with stealthchop. Not sure about the other 4.2.x boards, maybe someone who owns one can chime in.

Happy to try and answer any other questions you might have.
